Feat: CalFresh enrollment success with expiration

thekaveman opened the pull request:
Closes #1919 

## How to test

1. Temporarily change the [`context_processors.enrollment()`](https://github.com/cal-itp/benefits/blob/dev/benefits/core/context_processors.py#L75) to force a datetime for `expires`:

```diff
+ from datetime import datetime, timezone
...
- expiry = session.enrollment_expiry(request)
+ expiry = session.enrollment_expiry(request) or datetime.now(tz=timezone.utc)
```

2. In the Admin, update a flow that you can pass eligibility for to support expiration, e.g. MST Courtesy Cards
3. In the Admin, update the corresponding agency's PaymentProcessor to connect to the QA environment
4. Pass eligibility
5. Enroll a test credit card
6. See the new success page

## Screenshot - English

![image](https://github.com/cal-itp/benefits/assets/1783439/023c33a4-ab85-4f6d-a917-c546b0d892a9)

## Screenshot - Spanish

![image](https://github.com/cal-itp/benefits/assets/1783439/3254abc2-d3b6-46d3-bf28-4918e4f8f27f)

thekaveman commented:
I think we can actually merge this in, as it won't be activated for flows until we get back the expiration logic for enrollment (after #2052). And it doesn't have any conflicts with those changes, it just changed the `enrollment/success.html` template.

angela-tran reviewed (commented):
To test this locally, I: 
 - merged `dev` into my local `feat/calfresh-enrollment-success` branch so that I have the revert from #2052
 - edited my Courtesy Card eligibility type to support expiration
 - ran through the full Courtesy Card flow

I then see the success screen:
![image](https://github.com/cal-itp/benefits/assets/25497886/ab97f4e7-a972-49b8-bd41-8638d76bab71)

The expiration date doesn't show because currently there isn't any code that sets the `enrollment_expiry` session variable, and that's ok because this screen won't be accessible until after #2052 is reverted.

## Hard-coding in an `enrollment_expiry` to test date formatting
Just to test the date-formatting, I hard-coded a value in for `expiry` in the `session.enrollment_expiry` function:
```python
    # expiry = request.session.get(_ENROLLMENT_EXP)
    expiry = datetime(2024, 12, 1).timestamp()
```

English:

![image](https://github.com/cal-itp/benefits/assets/25497886/8ea7f44d-ef63-4761-a2c7-07838bad02c1)

Spanish:
![image](https://github.com/cal-itp/benefits/assets/25497886/50159bf6-aed5-4154-acab-e9fbae26b3d8)

@thekaveman Do you know how the date formatting gets set? I'm not sure why I'm getting an abbreviated month in English

angela-tran commented:
With a March date, I see the full month name:

![image](https://github.com/cal-itp/benefits/assets/25497886/fc68f905-03b6-4a1a-b8ad-469b076af82e)

thekaveman commented:
> @thekaveman Do you know how the date formatting gets set? I'm not sure why I'm getting an abbreviated month in English

We're using the [`date` filter](https://docs.djangoproject.com/en/5.0/ref/templates/builtins/#date), which uses the [`DATE_FORMAT`](https://docs.djangoproject.com/en/5.0/ref/settings/#date-format) setting by default, which is:

```python
DATE_FORMAT = "N j, Y"
```

Broken down as ([from the table here](https://docs.djangoproject.com/en/5.0/ref/templates/builtins/#date)):

* `N`: _Month abbreviation in Associated Press style. Proprietary extension._
* `j`: _Day of the month without leading zeros._
* `Y`: _Year, 4 digits with leading zeros._

It looks like for `N`, some months are fully spelled out and others are abbreviated. I guess we should change the default and instead use one of:

* `E`: _Month, locale specific alternative representation usually used for long date representation._
* `F`: _Month, textual, long._

I think each language has its own default format string, so changing this one should only affect English.

thekaveman commented:
@angela-tran I rebased on `dev` so this branch now has the revert from #2052.

I introduced custom date formats per locale and now we get the expected full-month formatting for both English and Spanish:

![image](https://github.com/cal-itp/benefits/assets/1783439/66c4093c-3d1c-418f-b04a-a9466f286e40)

![image](https://github.com/cal-itp/benefits/assets/1783439/06b73330-b0d2-4427-a6a9-092a7ba8127d)

angela-tran commented:
Nice!

angela-tran reviewed (commented):
The coverage report is flagging that the `formats.py` files aren't tested... should we do something about that?

thekaveman commented:
> The coverage report is flagging that the `formats.py` files aren't tested... should we do something about that?

Good call on this! I added tests for each in https://github.com/cal-itp/benefits/pull/1988/commits/9cdacdcf8e9768ec6e8e92895c57417195ad6be4

angela-tran reviewed (approved):
🚀

angela-tran commented:
Just noticed the infra pipeline failure... looking into it

angela-tran commented:
Re-ran it and it worked 😂
